nearcore/src/lib.rs

```diff
@@ -74,17 +71,87 @@ pub fn get_default_home() -> PathBuf {
     PathBuf::default()
 }
 
+fn db_checkpoint_root_dir<'a>(path: &'a Path, near_config: &'a NearConfig) -> &'a Path {
+    if let Some(db_checkpoints_path) = &near_config.config.db_checkpoints_path {
+        &db_checkpoints_path
+    } else {
+        path
+    }
+}
+
+const DB_CHECKPOINT_PREFIX: &str = "db_checkpoint_";
+
+fn find_db_checkpoint(path: &Path, near_config: &NearConfig) -> Result<Option<PathBuf>, io::Error> {
+    let db_checkpoints_path = db_checkpoint_root_dir(path, near_config);
+    for entry in std::fs::read_dir(db_checkpoints_path)?.filter(|entry| {
+        if let Ok(entry) = entry {
+            if let Some(file_name) = entry.file_name().to_str() {
+                return file_name.starts_with(DB_CHECKPOINT_PREFIX);
+            }
+        }
+        return false;
+    }) {
+        return Ok(Some(PathBuf::from(entry?.file_name())));
+    }
+    Ok(None)
+}
+
+fn create_db_checkpoint(path: &Path, near_config: &NearConfig) -> Result<PathBuf, anyhow::Error> {
+    let checkpoint_dir =
+        format!("{}{}", DB_CHECKPOINT_PREFIX, &chrono::Utc::now().format("%Y%m%d%H%M%S"));
+    let checkpoint_path = db_checkpoint_root_dir(path, near_config);
+    let checkpoint_path = checkpoint_path.join(Path::new(&checkpoint_dir));
+
+    let db = RocksDB::new(path)?;
+    let checkpoint = db.checkpoint()?;
+    info!(target:"near","Creating a DB snapshot in '{}'",checkpoint_path.display());
+    checkpoint.create_checkpoint(&checkpoint_path)?;
+
+    Ok(checkpoint_path)
+}
+
 /// Function checks current version of the database and applies migrations to the database.
 pub fn apply_store_migrations(path: &Path, near_config: &NearConfig) {
     let db_version = get_store_version(path);
     if db_version > near_primitives::version::DB_VERSION {
         error!(target: "near", "DB version {} is created by a newer version of neard, please update neard or delete data", db_version);
         std::process::exit(1);
     }
+
     if db_version == near_primitives::version::DB_VERSION {
         return;
     }
 
+    if near_config.config.use_checkpoints_for_db_migration {
+        let existing_checkpoint_path = find_db_checkpoint(path, near_config);
+        match existing_checkpoint_path {
+            Ok(existing_checkpoint_path) => match existing_checkpoint_path {
+                Some(existing_checkpoint_path) => {
+                    panic!("Detected an existing DB migration checkpoint: '{}'. If the previous DB migration attempt failed, please delete the checkpoint and restart. Otherwise, please restore from that checkpoint and restart.",existing_checkpoint_path.display());
```

Comment:
This can be inconvenient for users. What about using that snapshot?

Reply:
Technically, yes, a node can copy the files from a snapshot by itself. However, it needs to be done extremely carefully. I don't know a good way to separate files created by RocksDB vs user-created files. Thus, I can't assume that everything in ~/.near/data can be safely deleted.
Safe options of moving a checkpoint to '/tmp/' are not feasible, because '/tmp' can be a different filesystem.
As DB migrations are rare events, it should be ok to expect a user to recover manually.

An alternative to this PR is downloading a large public backup, which also requires a lot of manual work and a lot of time.